compiler/src/iree/compiler/Codegen/Common/GPU/GPULowerToGlobalLoads.cpp

```diff
@@ -0,0 +1,227 @@
+
+static constexpr int kNumBitsPerCopy = 32;
+
+static LogicalResult
+distributeLinalgCopyToThreads(RewriterBase &rewriter, linalg::CopyOp copy,
+                              ArrayRef<int64_t> workgroupSize,
+                              ArrayRef<int64_t> subgroupSize) {
+  LDBG("==== distributing op: ");
+  LDBG(*copy);
+  Location loc = copy.getLoc();
+
+  // The linalg.copy we are dealing with represents a region we need to copy to
+  // workgroup memory. Assume there are N threads in the workgroup, then there
+  // are `num_subgroups = N / gpu.subgroup_size` subgroups in the workgroup.
+  //
+  // So we are slicing up the target memref into `num_subgroups` consecutive
+  // slices, and threads in the same subgroup will copy their slice to workgroup
+  // memory slice.
+
+  // Get the copy size:
+  auto copyMemRefType = cast<MemRefType>(copy.getOperand(1).getType());
+  if (!memref::isStaticShapeAndContiguousRowMajor(copyMemRefType)) {
+    return rewriter.notifyMatchFailure(
+        copy, "Cannot proceed: copy to non-static or non-contiguous, "
+              "non-row major memref.");
+  }
+  auto rank = copyMemRefType.getRank();
+  SmallVector<OpFoldResult> tileSize(rank - 1, rewriter.getIndexAttr(1));
+
+  int64_t elementBitWidth = copyMemRefType.getElementTypeBitWidth();
+  if (kNumBitsPerCopy % elementBitWidth != 0) {
+    return rewriter.notifyMatchFailure(
+        copy, "Cannot proceed: preferred copy size is not a multiple of "
+              "element bit width.");
+  }
+  int64_t elementsPerCopy = kNumBitsPerCopy / elementBitWidth;
+
+  // Divide the copy by subgroup, and load linearly.
+  assert(subgroupSize.size() == 1); // only 1-D
+  assert(workgroupSize[0] % subgroupSize[0] == 0);
+
+  int64_t numSubgroups = workgroupSize[0] / subgroupSize[0];
+  int64_t totalCopySize = copyMemRefType.getNumElements();
+  int64_t totalCopySizePerSubgroup = totalCopySize / numSubgroups;
+  int64_t numCopiesPerThread =
+      (totalCopySizePerSubgroup / elementsPerCopy) / subgroupSize[0];
+  int64_t residualElements =
+      totalCopySizePerSubgroup % (subgroupSize[0] * elementsPerCopy);
+
+  LDBG("-- elementsPerCopy: " << elementsPerCopy);
+  LDBG("-- workgroupSize: " << workgroupSize[0]);
+  LDBG("-- numSubgroups: " << numSubgroups);
+  LDBG("-- totalCopySize: " << totalCopySize);
+  LDBG("-- totalCopySizePerSubgroup: " << totalCopySizePerSubgroup);
+  LDBG("-- numCopiesPerThread: " << numCopiesPerThread);
+  LDBG("-- residualElements: " << residualElements);
+
+  if (residualElements != 0) {
+    return rewriter.notifyMatchFailure(
+        copy, "Cannot proceed: cannot handle copying residual elements.");
+  }
+
+  Value subgroupId = rewriter.create<gpu::SubgroupIdOp>(loc, nullptr);
+  Value laneId = rewriter.create<gpu::LaneIdOp>(loc, nullptr);
+
+  auto sourceType = cast<MemRefType>(copy.getOperand(0).getType());
+  MemRefType localType = cast<MemRefType>(copy.getOutputs().front().getType());
```

Comment:
It is really hard to find the old conversion, so let me open a new one. (I could miss your response.)

nit: use auto because `cast` already spells the type. Also, we should be consistent here.

```suggestion
  auto sourceType = cast<MemRefType>(copy.getOperand(0).getType());
  auto localType = cast<MemRefType>(copy.getOutputs().front().getType());
```